### PR TITLE
Use tracing for logging

### PR DESCRIPTION
**Pull request by Dev380**
Attempts progress on #125

Currently, this PR switches logging to the tracing crate, and uses system loggers and tracing's stdout logger on mac and linux. I'm unable to test if it actually works on mac because I don't have access to one. Additionally, everything is instrumented with the #[instrument] macro so we have spans for free. Haven't figured out how to log to a system log storage yet on windows, and the linux implementation depends on systemd existing. Adding more logging to tun still needs to be completed.

### CODE REVIEW

File: burrow/src/main.rs [hunk 1]
```diff
@@ -27,17 +31,23 @@ enum Commands {
...
 
     Ok(())
 }
 
 #[tokio::main(flavor = "current_thread")]
 async fn main() {
+    LogTracer::init().expect("first thing in main - a logger shouldn't have been set yet");
+
+    let logger = init_logger_layer().with_subscriber(FmtSubscriber::new());
+    tracing::subscriber::set_global_default(logger).expect("Logger shouldn't be set already");
```
Comment (conradev) [inline burrow/src/main.rs:49]: Is there a way to configure tracing to print to stdout? perhaps with an environment variable?

**Review by conradev — CHANGES_REQUESTED** [review]
This looks fantastic! I requested a few changes

Do you think you could squash your commits into a single commit?

**Comment by conradev** [comment]
also, I can test on macOS!

File: burrow/src/main.rs (hunk 1 above)
Comment (Dev380) [reply burrow/src/main.rs]: I'm not sure, but AFAIK you need to add a logger for it to work. The FmtSubscriber logger should respect the RUST_LOG env variable

[the remark this answers is not in the document]
Comment (Dev380) [reply burrow/src/main.rs]: done

File: burrow/src/main.rs [hunk 2]
```diff
@@ -47,3 +57,13 @@ async fn main() {
         }
     }
 }
+
+#[cfg(target_os = "linux")]
+fn init_logger_layer() ->  tracing_journald::Layer {
+    tracing_journald::layer().expect("Couldn't open journald socket - are you using systemd?").with_syslog_identifier("burrow".to_string())
```
[the remark this answers is not in the document]
Comment (Dev380) [reply burrow/src/main.rs]: done

File: burrow/src/main.rs [hunk 3]
```diff
@@ -47,3 +57,13 @@ async fn main() {
         }
     }
 }
+
+#[cfg(target_os = "linux")]
+fn init_logger_layer() ->  tracing_journald::Layer {
```
[the remark this answers is not in the document]
Comment (Dev380) [reply burrow/src/main.rs]: done

File: burrow/src/main.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Dev380) [reply burrow/src/main.rs]: done - although I don't have a system/vm without journald available right now so I added a trace level log for later testing (it won't show up unless the RUST_LOG env variable is changed)

**Comment by Dev380** [comment]
> Do you think you could squash your commits into a single commit?

I'm not too familiar with git so I'm not sure how to do that - github says the repo owner has to [enable it](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/configuring-commit-squashing-for-pull-requests) because the option is grayed out for me
20230712_20h54m09s_grim

**Comment by conradev** [comment]
A git rebase is something you do locally, and is just a way of editing your commits. You'll have to do it locally, because Github doesn't help with it

I can show you how to do it when I get a chance, but also I squashed your commits for you in this case, so we should be good to merge

**Comment by conradev** [comment]
@Dev380 it looks like the builds are failing, though! You can ignore the "Apple Build" ones (which won't work because this is a PR from a fork and not a branch), but we need to fix the Rust Build errors

**Comment by conradev** [comment]
It looks like tracing-oslog is broken on some versions of Rust/macOS, which I might need to help with:
```
3 |         __dso_handle, _os_activity_create, _os_activity_current, mach_header,
  |                       ^^^^^^^^^^^^^^^^^^^  ^^^^^^^^^^^^^^^^^^^^ no `_os_activity_current` in `ffi`
```

It also looks like we need a manual `Debug` implementation on Windows for `TunInterface`

Linux looks like a typo and I can fix

**Comment by conradev** [comment]
Update: good to go, failing CI and merge conflicts

**Pull request closed without merging**